Make `MedianOfRunningArray` Generic

govardhanshah456 opened the pull request:
Closes #4390

govardhanshah456 commented:
@vil02 Hey, i did some changes according to my knowledge of generics. You can check it.

vil02 commented on src/main/java/com/thealgorithms/misc/MedianOfRunningArray.java:
```diff
@@ -38,14 +38,40 @@ public void insert(Integer e) {
-        if (maxHeap.size() == minHeap.size()) {
-            return (maxHeap.peek() + minHeap.peek()) / 2.0;
+    private T calculateAverage(T a, T b) {
+        if (a instanceof Integer) {
+            int sum = ((Integer) a) + ((Integer) b);
+            return (T) Integer.valueOf(sum / 2);
+        } else if (a instanceof Double) {
+            double sum = ((Double) a) + ((Double) b);
+            double roundedAverage = Math.round(sum / 2.0 * 100.0) / 100.0;
+            return (T) Double.valueOf(roundedAverage);
+        } else if (a instanceof Float) {
+            float sum = ((Float) a) + ((Float) b);
+            float roundedAverage = (float) (Math.round(sum / 2.0 * 100.0) / 100.0);
+            return (T) Float.valueOf(roundedAverage);
+        } else if (a instanceof Long) {
+            long sum = ((Long) a) + ((Long) b);
+            return (T) Long.valueOf(sum / 2);
+        } else if (a instanceof Short) {
+            int sum = ((Short) a) + ((Short) b);
+            return (T) Short.valueOf((short) (sum / 2));
+        } else if (a instanceof Byte) {
+            int sum = ((Byte) a) + ((Byte) b);
+            return (T) Byte.valueOf((byte) (sum / 2));
+        } else {
+            throw new IllegalArgumentException("Unsupported numeric type");
         }
-        return maxHeap.size() > minHeap.size() ? maxHeap.peek() * 1.0 : minHeap.peek() * 1.0;
     }
```
This is ugly. I think that it is fair to say that generics in Java are annoying. What do you think about defining the `MedianOfRunningArray` as abstract class (with `computeAverage` as an abstract method) and then have classes like `MedianOfRunningArrayInteger`, `MedianOfRunningArrayDouble` extending it?

In more details (I made some slight changes, please consider them):

```java
public abstract class MedianOfRunningArray<T extends Number & Comparable<T>> {

    private PriorityQueue<T> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
    private PriorityQueue<T> minHeap = new PriorityQueue<>();

    /*
      Inserting lower half of array to max Heap
      and upper half to min heap
     */
    public void insert(final T e) {
        if (!minHeap.isEmpty() && e.compareTo(minHeap.peek()) < 0) {
            maxHeap.offer(e);
            if (maxHeap.size() > minHeap.size() + 1) {
                minHeap.offer(maxHeap.poll());
            }
        } else {
            minHeap.offer(e);
            if (minHeap.size() > maxHeap.size() + 1) {
                maxHeap.offer(minHeap.poll());
            }
        }
    }

    /*
      Returns median at any given point
     */
    public T median() {
        if (maxHeap.isEmpty() && minHeap.isEmpty()) {
            throw new IllegalArgumentException("Enter at least 1 element, Median of empty list is not defined!");
        } else if (maxHeap.size() == minHeap.size()) {
            T maxHeapTop = maxHeap.peek();
            T minHeapTop = minHeap.peek();
            return calculateAverage(maxHeapTop, minHeapTop);
        }
        return maxHeap.size() > minHeap.size() ? maxHeap.peek() : minHeap.peek();
    }

    public abstract T calculateAverage(T a, T b);
}
```

and

```java
final public class MedianOfRunningArrayDouble extends MedianOfRunningArray<Double> {
    @Override
    public Double calculateAverage(final Double a, final Double b) {
        return (a + b) / 2.0d;
    }
}
```

```java
final public class MedianOfRunningArrayInteger extends MedianOfRunningArray<Integer> {
    @Override
    public Integer calculateAverage(final Integer a, final Integer b) {
        return (a + b) / 2;
    }
}
```

Having `calculateAverage` as public will help with testing.

The advantage of this approach is that when someone wants to have a different rounding for integers it is enough to have `MedianOfRunningArrayIntegerCrazyRounding`.

vil02 commented on src/test/java/com/thealgorithms/misc/MedianOfRunningArrayTest.java:
```diff
@@ -17,71 +17,71 @@ public class MedianOfRunningArrayTest {
 
```
Please clean-up the imports and the comment/documentation. I forgot to write that in the previous PR. Sorry.

vil02 replied:
`import static org.junit.jupiter.api.Assertions.assertFalse;` and `import org.apache.commons.lang3.tuple.Pair;` are not needed.

The comments
```
/**
 * Test case for Two sum Problem.
 * @author Bama Charan Chhandogi (https://github.com/BamaCharanChhandogi)
 */
```
seems to be wrong.

govardhanshah456 commented:
@vil02 Hey ,as suggested I have made class abstract and did the required changes, have a look.

vil02 commented on src/main/java/com/thealgorithms/misc/MedianOfRunningArrayInteger.java:
```diff
@@ -0,0 +1,8 @@
+package com.thealgorithms.misc;
+
+final public class MedianOfRunningArrayInteger extends MedianOfRunningArray<Integer> {
```
```suggestion
public final class MedianOfRunningArrayInteger extends MedianOfRunningArray<Integer> {
```

vil02 commented on src/main/java/com/thealgorithms/misc/MedianOfRunningArrayLong.java:
```diff
@@ -0,0 +1,8 @@
+package com.thealgorithms.misc;
+
+final public class MedianOfRunningArrayLong extends MedianOfRunningArray<Long> {
```
```suggestion
public final class MedianOfRunningArrayLong extends MedianOfRunningArray<Long> {
```

vil02 reviewed (changes requested):
It goes in a very good direction!

govardhanshah456 commented:
@vil02 Hey, modified that comments and removed that impot and committed suggested changes, regarding double  and float i did not know that there exist some way to check precision using assertEqual itself, so was formatting at output side.

vil02 reviewed (dismissed):
Looks good. Thanks!

@siriak could you please have a look? There were _lots_ of new files introduced and I would like to double check if the current structure is correct.

vil02 commented on src/main/java/com/thealgorithms/misc/MedianOfRunningArrayFloat.java:
```diff
@@ -0,0 +1,7 @@
+package com.thealgorithms.misc;
+public final class MedianOfRunningArrayFloat extends MedianOfRunningArray<Float> {
```
```suggestion

public final class MedianOfRunningArrayFloat extends MedianOfRunningArray<Float> {
```

govardhanshah456 commented:
Yeah, thanks.

siriak reviewed (approved):
Looks good, thanks!

vil02 commented:
@govardhanshah456 this was a good piece of Java programing. Thanks!
